app/src/main/java/org/schabi/newpipe/player/seekbarpreview/SeekbarPreviewThumbnailHolder.java

```diff
@@ -0,0 +1,252 @@
+                    .orElse(null);
+        } else {
+            Log.d(TAG, "Strategy for seekbarPreviewData: low quality");
+            return framesets.stream()
+                    .min(Comparator.comparingInt(fs -> fs.getFrameHeight() * fs.getFrameWidth()))
+                    .orElse(null);
+        }
+    }
+
+    private void generateDataFrom(
+            final Frameset frameset,
+            final UUID updateRequestIdentifier) {
+
+        Log.d(TAG, "Starting generation of seekbarPreviewData");
+        final Stopwatch sw = Log.isLoggable(TAG, Log.DEBUG) ? Stopwatch.createStarted() : null;
+
+        int currentPosMs = 0;
+        int pos = 1;
+
+        final int frameCountPerUrl = frameset.getFramesPerPageX() * frameset.getFramesPerPageY();
+
+        // Process each url in the frameset
+        for (final String url : frameset.getUrls()) {
+            // get the bitmap
+            final Bitmap srcBitMap = getBitMapFrom(url);
+
+            // The data is not added directly to "seekbarPreviewData" due to
+            // concurrency and checks for "updateRequestIdentifier"
+            final Map<Integer, Supplier<Bitmap>> generatedDataForUrl = new HashMap<>();
+
+            // The bitmap consists of several images, which we process here
+            // foreach frame in the returned bitmap
+            for (int i = 0; i < frameCountPerUrl; i++) {
+                // Frames outside the video length are skipped
+                if (pos > frameset.getTotalCount()) {
+                    break;
+                }
+
+                // Get the bounds where the frame is found
+                final int[] bounds = frameset.getFrameBoundsAt(currentPosMs);
+                generatedDataForUrl.put(currentPosMs, () -> {
+                    // It can happen, that the original bitmap could not be downloaded
+                    // In such a case - we don't want a NullPointer - simply return null
+                    if (srcBitMap == null) {
+                        return null;
+                    }
+
+                    // Cut out the corresponding bitmap form the "srcBitMap"
+                    return Bitmap.createBitmap(srcBitMap, bounds[1], bounds[2],
+                            frameset.getFrameWidth(), frameset.getFrameHeight());
+                });
+
+                currentPosMs += frameset.getDurationPerFrame();
+                pos++;
+            }
+
+            // Check if we are still the latest request
+            // If not abort method execution
+            if (isRequestIdentifierCurrent(updateRequestIdentifier)) {
+                seekbarPreviewData.putAll(generatedDataForUrl);
+            } else {
+                Log.d(TAG, "Aborted of generation of seekbarPreviewData");
+                break;
+            }
+        }
+
+        if (sw != null) {
+            Log.d(TAG, "Generation of seekbarPreviewData took " + sw.stop().toString());
```

Comment:
At first I was questioning the decision to keep this stopwatch code, but I guess it is just useful and should be kept.

Reply:
Can I resolve this conversation or is this a note for other reviewers and I should leave it open?